Tree Based Models - Q12 - 14/July
===================================
Data pertaining the startup companies is provided in 05_Startups_profit.csv

https://drive.google.com/drive/folders/1Jl8iDu7nGmrqCECbrLqmVafgwE5PYfiU

Train a decision tree to predict the profit of the companies based on other variables. 

Does the state in which the startup is , impact its profit?

In [1]:
# Import libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.tree import DecisionTreeRegressor, plot_tree, export_text

import warnings
warnings.filterwarnings("ignore")

In [2]:
# Read and display the data file
startups_data = pd.read_csv('/Users/riteshturlapaty/ai-ml-learning/AccelerateAI/7.DecisionTree/DailyQuiz/05_Startups_profit.csv')
startups_data.head(5)

R&D Spend  Administration  Marketing Spend       State     Profit
0  165349.20       136897.80        471784.10    New York  192261.83
1  162597.70       151377.59        443898.53  California  191792.06
2  153441.51       101145.55        407934.54     Florida  191050.39
3  144372.41       118671.85        383199.62    New York  182901.99
4  142107.34        91391.77        366168.42     Florida  166187.94

In [3]:
startups_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   R&D Spend        50 non-null     float64
 1   Administration   50 non-null     float64
 2   Marketing Spend  50 non-null     float64
 3   State            50 non-null     object 
 4   Profit           50 non-null     float64
dtypes: float64(4), object(1)
memory usage: 2.1+ KB


In [4]:
# Create dummies for categorical variable State
startup_dummies = pd.get_dummies(startups_data, columns=['State'])
startup_dummies.sample(5)

R&D Spend  Administration  Marketing Spend     Profit  State_California  \
6   134615.46       147198.87        127716.82  156122.51                 1   
11  100671.96        91790.61        249744.55  144259.40                 1   
37   44069.95        51283.14        197029.42   89949.14                 1   
28   66051.52       182645.56        118148.20  103282.38                 0   
19   86419.70       153514.11             0.00  122776.86                 0   

    State_Florida  State_New York  
6               0               0  
11              0               0  
37              0               0  
28              1               0  
19              0               1

In [5]:
# Train the model usin CART Regression Tree
X = startup_dummies.drop('Profit', axis=1)
y = startup_dummies['Profit']

clf = DecisionTreeRegressor(random_state=42)
clf = clf.fit(X,y)

In [6]:
# Calculate Accuracy
clf.score(X,y)

1.0

In [7]:
# Get the feature importances
pd.Series(clf.feature_importances_, index=X.columns)

R&D Spend           0.866580
Administration      0.001761
Marketing Spend     0.130431
State_California    0.001160
State_Florida       0.000054
State_New York      0.000014
dtype: float64

As seen. below, feature importance of state attribute is having very marginal values and hence it can be concluded that state in which startups are founded, hardly make any difference

State_California    0.001160

State_Florida       0.000054

State_New York      0.000014